cmd/nerdctl/compose_ps.go

```diff
@@ -111,3 +125,114 @@ func composePsAction(cmd *cobra.Command, args []string) error {
 
 	return w.Flush()
 }
+
+// composeContainerPrintableTab constructs composeContainerPrintable with fields
+// only for console output.
+func composeContainerPrintableTab(ctx context.Context, container containerd.Container) (composeContainerPrintable, error) {
+	info, err := container.Info(ctx, containerd.WithoutRefreshedMetadata)
+	if err != nil {
+		return composeContainerPrintable{}, err
+	}
+	spec, err := container.Spec(ctx)
+	if err != nil {
+		return composeContainerPrintable{}, err
+	}
+	status := formatter.ContainerStatus(ctx, container)
+	if status == "Up" {
+		status = "running" // corresponds to Docker Compose v2.0.1
+	}
+
+	return composeContainerPrintable{
+		Name:    info.Labels[labels.Name],
+		Command: formatter.InspectContainerCommandTrunc(spec),
+		Service: info.Labels[labels.ComposeService],
+		State:   status,
+		Ports:   formatter.FormatPorts(info.Labels),
+	}, nil
+}
+
+// composeContainerPrintableTab constructs composeContainerPrintable with fields
+// only for json output and compatible docker output.
+func composeContainerPrintableJSON(ctx context.Context, container containerd.Container) (composeContainerPrintable, error) {
+	info, err := container.Info(ctx, containerd.WithoutRefreshedMetadata)
+	if err != nil {
+		return composeContainerPrintable{}, err
+	}
+	spec, err := container.Spec(ctx)
+	if err != nil {
+		return composeContainerPrintable{}, err
+	}
+
+	var (
+		state    string
+		exitCode uint32
+	)
+	status, err := containerStatus(ctx, container)
+	if err == nil {
+		// show exitCode only when container is exited/stopped
+		if status.Status == containerd.Stopped {
+			exitCode = status.ExitStatus
+		}
+		state = string(status.Status)
+	} else {
+		state = string(containerd.Unknown)
+	}
+
+	return composeContainerPrintable{
+		ID:         container.ID(),
+		Name:       info.Labels[labels.Name],
+		Command:    formatter.InspectContainerCommand(spec, false, false),
+		Project:    info.Labels[labels.ComposeProject],
+		Service:    info.Labels[labels.ComposeService],
+		State:      state,
+		Health:     "",
+		ExitCode:   exitCode,
+		Publishers: formatPublishers(info.Labels),
+	}, nil
+}
+
+func containerStatus(ctx context.Context, c containerd.Container) (containerd.Status, error) {
+	// Just in case, there is something wrong in server.
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	task, err := c.Task(ctx, nil)
+	if err != nil {
+		return containerd.Status{}, err
+	}
+
+	return task.Status(ctx)
+}
+
+// PortPublisher hold status about published port
```

Comment:
I copied this struct from docker to avoid importing a whole package. Put it (and the below `formatPublishers` func) here since they are only relevant to `compose ps --format json`.